feat(loader): The loader component. Full size and inline.

keepitterron opened the pull request:
![loader](https://user-images.githubusercontent.com/593897/28875484-043afbe2-7796-11e7-8d5e-dd8cb0e6e2c8.gif)

dine commented on src/index.scss:
```diff
@@ -58,6 +58,7 @@ $ns: 'sb-' !default;
+@import 'components/loader';
 
-// Trumps
+// Trumps @>:^0
```
🤡

dine commented on src/components/_loader.scss:
```diff
@@ -0,0 +1,111 @@
+// Loader
+//
+// Markup:
+// <div class="sb-loader sb-loader--inline">
```
If we can make the markup the same and the only change is the modifier class, we could add `.sb-loader--inline`declared after the markup, like we do in `buttons.scss` for example.

dine commented on src/components/_loader.scss:
```diff
@@ -0,0 +1,111 @@
+  flex-direction: column;
+  align-items: center;
+
+  &-label {
```
We are using [BEM](http://getbem.com/naming/), so the elements should have a `__` instead of `-`.

dine commented on src/components/_loader.scss:
```diff
@@ -0,0 +1,111 @@
+    padding: 0 $sb-loader-padding-inline;
+
+    .#{$ns}loader-spinner {
+      margin-left: auto;
```
If we use the approach of same markup with modifier class, here we will have to declare here the new size / position for the spinner.

endritbajo commented on src/components/_loader.scss:
```diff
@@ -0,0 +1,114 @@
+    background: $sb-color-white;
+    border: 1px solid $sb-color-light-3;
+    width: $sb-loader-size;
+    border-radius: 2px;
```
Here we can use `$sb-radius`

endritbajo commented on src/components/_loader.scss:
```diff
@@ -0,0 +1,114 @@
+$sb-loader-animation-duration: 0.8s;
+
+.#{$ns}loader {
+  display: flex;
```
What about browser support, does it work on IE < 11?

beerecca commented on src/components/_loader.scss:
```diff
@@ -0,0 +1,114 @@
+// Markup:
+// <div class="sb-loader sb-loader--inline">
+//   <span class="sb-loader__label">Loading</span>
+//   <svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="51" height="50" viewBox="0 0 51 50" class="sb-loader__spinner">
```
Lots we can do to tidy up this svg. Take a look at https://jakearchibald.github.io/svgomg/ to start with (once I guess we get a new version from Gal).

beerecca replied:
Check accessibility on this, especially with a screen reader - should at least have aria-hidden on the loader I think?

beerecca replied:
Should remove the fixed width and height from the svg too.

beerecca commented on src/components/_loader.scss:
```diff
@@ -0,0 +1,114 @@
+    animation-timing-function: linear;
+  }
+
+  &__spinner--primary {
```
This isn't a modifier so shouldn't have `--`. Can instead create a new base - `sb-spinner` (not nested inside the loader class), with `sb-spinner__primary` and `sb-spinner__secondary`.

beerecca commented on src/components/_loader.scss:
```diff
@@ -0,0 +1,114 @@
+  align-items: center;
+
+  &__label {
+    padding: $sb-loader-padding-inline 0;
```
Can we not use an existing padding class? Same below

keepitterron replied:
I couldn't find a `13px` spacing generic mixin/class used elsewhere?

beerecca commented on src/components/_loader.scss:
```diff
@@ -0,0 +1,114 @@
+//
+// Styleguide 16.
+
+$sb-loader-size: sb-px2rems(200px);
```
Same issue we had with the dropdown - I don't think we should fix this. I think it should be 100%, then the user can put it in whatever container (or grid column) they want. (If we do stick with this though it should be `sb-inline-loader-width`)

keepitterron replied:
I'll do a style check with Gal tomorrow. I also think this should be just 100% width.

keepitterron commented:
Did a major refactoring after getting more familiar with the tool and processes and reading through your comments.
I also had to recreate from scratch the svg spinner.
Also rewrote git history by rebasing and removing obsolete commits, sorry about that :P

keepitterron commented:
I should have addressed all your comments, if not pls bring it up.
- old browser support
- aria-hidden
- separate `sb-spinner` class
- proper use of BEM
- same markup
- smooth, optimized svg

dine commented on src/index.scss:
```diff
@@ -58,6 +58,7 @@ $ns: 'sb-' !default;
 @import 'components/select';
+@import 'components/loader';
 
 // Trumps
```
no more  @>:^0? ☹️

keepitterron replied:
oh nooooo, i lost it with `git rebase` :(

beerecca commented on src/components/_loader.scss:
```diff
@@ -0,0 +1,89 @@
+//
+// Markup:
+// <div class="sb-loader {{modifier_class}}">
+//   <svg viewBox="0 0 50 50" class="sb-spinner" aria-hidden="false">
```
Should be aria hidden true. Because it won't mean anything to a blind person. Only the loading label will. Maybe do a bit of research on this, what's the best approach for a spinner?

beerecca replied:
Still lots that can be cleaned up with the svg. Shouldn't need all those ids, groups, stroke rules, widths and heights etc.

keepitterron replied:
already on it :)

beerecca commented on src/components/_loader.scss:
```diff
@@ -0,0 +1,89 @@
+//
+// Styleguide 16.
+
+$sb-loader-size--inline: sb-px2rems(200px);
```
Want to change this before we merge it

beerecca replied:
Want to change this before we merge it

beerecca commented on src/components/_loader.scss:
```diff
@@ -0,0 +1,89 @@
+$sb-loader-size--inline: sb-px2rems(200px);
+$sb-loader-spinner-size: sb-px2rems(50px);
+$sb-loader-spinner-size--inline: sb-px2rems(16px);
+$sb-loader-padding: sb-px2rems(13px);
```
Couldn't comment on your comment for some reason  - but I really don't like the idea of a random new padding. Will one of the others still look good?

keepitterron replied:
yes, it's on my list of things to discuss with gal. our padding small is `11px`, don't see much difference there.

keepitterron commented:
Last update.
- SVG only has properties it really needs
- did a spacing check with Gal, using 11px/16px now
- parent will decide the inline loader size (100%)
